File: Leetcode/MakeSumDivisiblebyP.rs
// https://leetcode.com/problems/make-sum-divisible-by-p

use std::collections::HashMap;

struct Solution {}

impl Solution {
    pub fn min_subarray(nums: Vec<i32>, p: i32) -> i32 {
        let mut nums = nums.into_iter().map(|x| x as usize).collect::<Vec<_>>();
        nums.push(0);
        let p = p as usize;
        let n = nums.len();

        let mut rems = HashMap::<usize, usize>::new();
        rems.insert(0, 0);

        let mut used = 0;
        let mut rest = nums.iter().sum::<usize>();
        let mut ans = 0;

        for i in 0..n {
            let rem = rest % p;
            let need = (p - rem) % p;
            if let Some(&pos) = rems.get(&need) {
                ans = ans.max(n - i + pos);
            }

            used += nums[i];
            rest -= nums[i];
            rems.insert(used % p, i + 1);
        }

        let ans = n - ans;
        if ans == n - 1 {
            -1
        } else {
            ans as i32
        }
    }
}

File: Leetcode/MakeSumDivisiblebyP.cpp
// https://leetcode.com/problems/make-sum-divisible-by-p

class Solution {
   public:
    int minSubarray(vector<int>& nums, int p) {
        using ll = long long;

        const size_t n = nums.size();

        ll sum = 0;
        for (size_t i = 0; i < n; ++i) {
            sum += nums[i];
        }

        map<ll, size_t> max_pref;
        max_pref[0] = 0;

        ll pref = 0;
        size_t ans = n;
        for (size_t i = 0; i < n; ++i) {
            pref += nums[i];

            const ll rest = sum - pref;
            const ll extra = rest % p;

            if (pref % p == 0) {
                ans = min(ans, n - i - 1);
            }
            if (max_pref.cend() != max_pref.find((p - extra) % p)) {
                ans = min(ans, i + 1 - max_pref[(p - extra) % p]);
            }

            max_pref[pref % p] = i + 1;
        }

        return ans == n ? -1 : ans;
    }
};